File: typescript/examples/object-pick.test.ts
// カード object-pick の Contract を検証するテスト
import { pick } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

describe('object-pick: pick', () => {
  it('指定したキーだけを持つ新しいオブジェクトを返す', () => {
    const user = { id: 1, name: 'a', password: 'x' };
    expect(pick(user, ['id', 'name'])).toEqual({ id: 1, name: 'a' });
  });

  it('入力を変更せず、ネストした値は同じ参照（浅いコピー）', () => {
    const nested = { x: 1 };
    const input = { n: nested, other: 2 };
    const result = pick(input, ['n']);
    expect(input).toEqual({ n: { x: 1 }, other: 2 });
    expect(result).not.toBe(input);
    expect(result.n).toBe(nested);
  });

  it('自身のプロパティだけを対象にし、継承したプロパティは取り出せない', () => {
    const obj: Record<string, number> = Object.create({ inherited: 1 });
    obj.own = 2;
    expect(pick(obj, ['inherited', 'own'])).toEqual({ own: 2 });
  });

  it('存在しないキーは無視し、undefined を持つ自身のキーは含める', () => {
    const obj: Record<string, number | undefined> = { a: 1, u: undefined };
    const result = pick(obj, ['a', 'missing', 'u']);
    expect(Object.hasOwn(result, 'missing')).toBe(false);
    expect(Object.hasOwn(result, 'u')).toBe(true);
    expect(result).toEqual({ a: 1, u: undefined });
  });

  it('返り値のキー順は keys の並び順で、重複は 1 つにまとまる', () => {
    expect(Object.keys(pick({ a: 1, b: 2, c: 3 }, ['c', 'a', 'c']))).toEqual(['c', 'a']);
  });

  it('keys が空なら空オブジェクトを返す', () => {
    expect(pick({ a: 1 }, [])).toEqual({});
  });
});

File: python/examples/object-pick_test.py
# カード object-pick の Contract を検証するテスト
import operator

import pytest


def pick(d: dict, keys) -> dict:
    return {k: d[k] for k in keys if k in d}


def test_picks_only_given_keys():
    """指定したキーだけを持つ辞書を返す"""
    user = {"id": 1, "name": "a", "password": "x"}
    assert pick(user, ["id", "name"]) == {"id": 1, "name": "a"}


def test_does_not_mutate_input_and_returns_shallow_copy():
    """入力辞書を変更せず、値は浅いコピー（ネストした辞書は同じ参照）"""
    nested = {"z": 1}
    user = {"id": 1, "nested": nested}
    result = pick(user, ["nested"])
    assert user == {"id": 1, "nested": nested}
    assert result is not user
    assert result["nested"] is nested


def test_missing_keys_are_ignored_and_none_values_are_kept():
    """存在しないキーは無視され、値が None のキーは None のまま含まれる"""
    d = {"a": None, "b": 2}
    assert pick(d, ["a", "missing"]) == {"a": None}
    assert "missing" not in pick(d, ["missing"])


def test_result_follows_keys_order_and_dedupes():
    """返り値のキーは keys の順で、重複したキーは 1 つにまとまる"""
    d = {"a": 1, "b": 2, "c": 3}
    result = pick(d, ["c", "a", "c"])
    assert list(result) == ["c", "a"]
    assert result == {"c": 3, "a": 1}


def test_empty_keys_returns_empty_dict():
    """keys が空なら {} を返す"""
    assert pick({"a": 1}, []) == {}


def test_unhashable_key_raises_type_error():
    """ハッシュ化できないキーがあると TypeError を投げる"""
    with pytest.raises(TypeError):
        pick({"a": 1}, [["a"]])


def test_itemgetter_raises_key_error_for_missing_key():
    """Alternatives: operator.itemgetter は存在しないキーで KeyError になる"""
    d = {"id": 1, "name": "a"}
    assert operator.itemgetter("id", "name")(d) == (1, "a")
    with pytest.raises(KeyError):
        operator.itemgetter("id", "missing")(d)


def test_get_variant_includes_missing_keys_as_none():
    """Pitfalls: d.get(k) で書くと存在しないキーが None で含まれる"""
    d = {"a": 1}
    assert {k: d.get(k) for k in ["a", "missing"]} == {"a": 1, "missing": None}
